com.unity.netcode.gameobjects/Runtime/Components/Helpers/ComponentController.cs

```diff
@@ -0,0 +1,543 @@
+        [SerializeField]
+        internal List<ComponentEntry> Components;
+
+        /// <summary>
+        /// Returns the current enabled state of the <see cref="ComponentController"/>.
+        /// </summary>
+        public bool EnabledState => m_IsEnabled;
+
+        internal List<ComponentEntry> ValidComponents = new List<ComponentEntry>();
+        private bool m_IsEnabled;
+
+#if UNITY_EDITOR
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool IsValidComponentType(Object component)
+        {
+            return !(component.GetType().IsSubclassOf(typeof(NetworkBehaviour)) || component.GetType() == typeof(NetworkObject) || component.GetType() == typeof(NetworkManager));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static string GetComponentNameFormatted(Object component)
+        {
+            // Split the class name up based on capitalization
+            var classNameDisplay = Regex.Replace(component.GetType().Name, "([A-Z])", " $1", RegexOptions.Compiled).Trim();
+            return $"{component.name} ({classNameDisplay})";
+        }
+
+        /// <inheritdoc/>
+        /// <remarks>
+        /// Checks for invalid <see cref="Object"/> entries.
+        /// </remarks>
+        protected virtual void OnValidate()
+        {
+            if (Components == null || Components.Count == 0)
+            {
+                return;
+            }
+
+            var gameObjectsToScan = new List<ComponentEntry>();
+            // First pass is to verify all entries are valid and look for any GameObjects added as an entry to process next
+            for (int i = Components.Count - 1; i >= 0; i--)
+            {
+                if (Components[i] == null)
+                {
+                    continue;
+                }
+
+                if (Components[i].Component == null)
+                {
+                    continue;
+                }
+                var objectType = Components[i].Component.GetType();
+                if (objectType == typeof(GameObject))
+                {
+                    if (!gameObjectsToScan.Contains(Components[i]))
+                    {
+                        gameObjectsToScan.Add(Components[i]);
+                    }
+                    Components.RemoveAt(i);
+                    continue;
+                }
+
+                if (!IsValidComponentType(Components[i].Component))
+                {
+                    Debug.LogWarning($"Removing {GetComponentNameFormatted(Components[i].Component)} since {Components[i].Component.GetType().Name} is not an allowed component type.");
+                    Components.RemoveAt(i);
+                    continue;
+                }
+
+                var propertyInfo = Components[i].Component.GetType().GetProperty("enabled", BindingFlags.Instance | BindingFlags.Public);
+                if (propertyInfo == null && propertyInfo.PropertyType != typeof(bool))
```

Comment:
I'm guessing we don't want to check for `is null && has property`?

Reply:
Actually... that should be:
`if (propertyInfo == null || propertyInfo.PropertyType != typeof(bool))`

Thank you for commenting on that one! 
🥇